codec-http2/src/main/java/io/netty/handler/codec/http2/internal/hpack/HuffmanDecoder.java

```diff
@@ -213,4 +138,114 @@ private static void insert(Node root, int symbol, int code, byte length) {
+         * of the parent nodes. symbolBits tells how far down the tree we are. For example, when
+         * decoding the invalid sequence {0xff, 0xff}, currentBits will be 0, but symbolBits will be
+         * 16. This is used to know if buf ended early (before consuming a whole symbol) or if
+         * there is too much padding.
+         */
+        @Override
+        public boolean process(byte value) throws IOException {
+            int b = value & 0xFF;
+            current = (current << 8) | b;
+            currentBits += 8;
+            symbolBits += 8;
+            // While there are unconsumed bits in current, keep consuming symbols.
+            while (currentBits >= 8) {
+                int c = (current >>> (currentBits - 8)) & 0xFF;
+                node = node.children[c];
+                currentBits -= node.bits;
+                if (node.isTerminal()) {
+                    if (node.symbol == HpackUtil.HUFFMAN_EOS) {
+                        throw EOS_DECODED;
+                    }
+                    append(node.symbol);
+                    node = ROOT;
+                    // Upon consuming a whole symbol, reset the symbol bits to the number of bits
+                    // left over in the byte.
+                    symbolBits = currentBits;
+                }
+            }
+            return true;
+        }
+
+        AsciiString end() throws IOException {
+            /*
+             * We have consumed all the bytes in buf, but haven't consumed all the symbols. We may be on
+             * a partial symbol, so consume until there is nothing left. This will loop at most 2 times.
+             */
+            while (currentBits > 0) {
+                int c = (current << (8 - currentBits)) & 0xFF;
+                node = node.children[c];
+                if (node.isTerminal() && node.bits <= currentBits) {
+                    if (node.symbol == HpackUtil.HUFFMAN_EOS) {
+                        throw EOS_DECODED;
+                    }
+                    currentBits -= node.bits;
+                    append(node.symbol);
+                    node = ROOT;
+                    symbolBits = currentBits;
+                } else {
+                    break;
+                }
+            }
+
+            // Section 5.2. String Literal Representation
+            // A padding strictly longer than 7 bits MUST be treated as a decoding error.
+            // Padding not corresponding to the most significant bits of the code
+            // for the EOS symbol (0xFF) MUST be treated as a decoding error.
+            int mask = (1 << symbolBits) - 1;
+            if (symbolBits > 7 || (current & mask) != mask) {
+                throw INVALID_PADDING;
+            }
+
+            return new AsciiString(bytes, 0, index, false);
+        }
+
+        private void append(int i) {
+            try {
+                bytes[index] = (byte) i;
+            } catch (IndexOutOfBoundsException ignore) {
```

Comment:
@nmittler - I experimented with this when updating [AppendableCharSequence](https://github.com/netty/netty/blob/4.1/common/src/main/java/io/netty/util/internal/AppendableCharSequence.java) and it did seem to provide performance benefit assuming the majority of the time you will not exceed the array length (see [AppendableCharSequenceBenchmark](https://github.com/netty/netty/blob/4.1/microbench/src/main/java/io/netty/microbenchmark/common/AppendableCharSequenceBenchmark.java)) ... having a good guess of array length can also really help here.